### ISSUE TITLE
MongoDB `upsert()`: `Internal error: Attempted to serialize empty result.`

### ISSUE THREAD

**Issue by khufu-adecoster** [prisma/prisma#20527]
Hi Prisma Team! My Prisma Client just crashed. This is the report:
## Versions

| Name            | Version            |
|-----------------|--------------------|
| Node            | v16.14.0           | 
| OS              | darwin             |
| Prisma Client   | 5.0.0              |
| Query Engine    | 6b0aef69b7cdfc787f822ecd7cdc76d5f1991584|
| Database        | mongodb            |

## Logs
```
prisma:tryLoadEnv Environment variables loaded from /Users/adecoster/code/projects/clokizi/api/.env
prisma:client checkPlatformCaching:postinstall false
prisma:client checkPlatformCaching:ciName 
prisma:tryLoadEnv Environment variables loaded from /Users/adecoster/code/projects/clokizi/api/.env
prisma:client dirname /Users/adecoster/code/projects/clokizi/api/node_modules/.prisma/client
prisma:client relativePath ../../../prisma
prisma:client cwd /Users/adecoster/code/projects/clokizi/api/prisma
prisma:client clientVersion 5.0.0
prisma:client clientEngineType library
prisma:client:libraryEngine internalSetup
prisma:client:engines:resolveEnginePath enginePath /Users/adecoster/code/projects/clokizi/api/node_modules/.prisma/client/libquery_engine-darwin.dylib.node
prisma:client:libraryEngine sending request, this.libraryStarted: false
prisma:client:libraryEngine library starting
prisma:client:libraryEngine library started
```

## Client Snippet
```ts
import { PrismaClient } from '@prisma/client'

const globalForPrisma = globalThis as unknown as {
  prisma: PrismaClient | undefined
}

export const prisma = globalForPrisma.prisma ?? new PrismaClient()

if (process.env.NODE_ENV !== 'production') globalForPrisma.prisma = prisma
```

## Schema
```prisma
generator client {
  provider        = "prisma-client-js"
  previewFeatures = []
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model Users {
  //ids
  id                    String           @id @default(cuid()) @map("_id")
  firebaseId            String           @unique
  //informations
  name                  String?
  email                 String?
  phone                 String?
  language              Languages?
  avatar                String?
  companies             UsersCompanies[] @relation("UsersCompaniesUser")
  //request
  lastRequestAt         DateTime?
  //create
  createdAt             DateTime         @default(now())
  createdById           String?
  createdBy             Users?           @relation("UsersCreatedBy", fields: [createdById], references: [id], onDelete: NoAction, onUpdate: NoAction)
  //update
  updatedAt             DateTime?        @updatedAt
  updatedById           String?
  updatedBy             Users?           @relation("UsersUpdatedBy", fields: [updatedById], references: [id], onDelete: NoAction, onUpdate: NoAction)
  //fk users
  createdUsers          Users[]          @relation("UsersCreatedBy")
  updatedUsers          Users[]          @relation("UsersUpdatedBy")
  //fk companies
  createdCompanies      Companies[]      @relation("CompaniesCreatedBy")
  updatedCompanies      Companies[]      @relation("CompaniesUpdatedBy")
  //fk users <-> companies
  createdUsersCompanies UsersCompanies[] @relation("UsersCompaniesCreatedBy")
  updatedUsersCompanies UsersCompanies[] @relation("UsersCompaniesUpdatedBy")
  //fk customers
  createdCustomers      Customers[]      @relation("CustomersCreatedBy")
  updatedCustomers      Customers[]      @relation("CustomersUpdatedBy")
  //fk invoices
  createdInvoices       Invoices[]       @relation("InvoicesCreatedBy")
  updatedInvoices       Invoices[]       @relation("InvoicesUpdatedBy")
}

model Companies {
  //ids
  id          String           @id @default(cuid()) @map("_id")
  //informations
  name        String
  avatar      String?
  country     Countries?
  users       UsersCompanies[] @relation("UsersCompaniesCompany")
  customers   Customers[]      @relation("CustomersCompany")
  //create
  createdAt   DateTime         @default(now())
  createdById String?
  createdBy   Users?           @relation("CompaniesCreatedBy", fields: [createdById], references: [id], onDelete: NoAction, onUpdate: NoAction)
  //update
  updatedAt   DateTime?        @updatedAt
  updatedById String?
  updatedBy   Users?           @relation("CompaniesUpdatedBy", fields: [updatedById], references: [id], onDelete: NoAction, onUpdate: NoAction)
}

model UsersCompanies {
  id          String    @id @default(cuid()) @map("_id")
  //informations 
  role        Roles
  companyId   String
  company     Companies @relation("UsersCompaniesCompany", fields: [companyId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  userId      String
  user        Users     @relation("UsersCompaniesUser", fields: [userId], references: [id], onDelete: NoAction, onUpdate: NoAction)
  //create
  createdAt   DateTime  @default(now())
  createdById String?
  createdBy   Users?    @relation("UsersCompaniesCreatedBy", fields: [createdById], references: [id], onDelete: NoAction, onUpdate: NoAction)
  //update
  updatedAt   DateTime? @updatedAt
  updatedById String?
  updatedBy   Users?    @relation("UsersCompaniesUpdatedBy", fields: [updatedById], references: [id], onDelete: NoAction, onUpdate: NoAction)
}

enum Roles {
  ADMIN
  MANAGER
  ACCOUNTANT
  AGENT
}

enum Languages {
  EN
  ES
  FR
}
```

## Prisma Engine Query
```
import { prisma } from '../db'

let user: Prisma.UsersGetPayload<{ include: { companies: true } }> = await prisma.users.upsert({
    where: {
      firebaseId: uid
    },
    create: {
      firebaseId: uid,
      email,
      name
    },
    update: {
      // lastRequestAt: new Date(),
      email,
      phone
    },
    include: {
      companies: true
    }
  })
```

**Comment by janpio** [comment]
Thanks for opening an issue @khufu-adecoster. Is this a common occurrence or was this a one-off thing that happened out of the blue?

Was this a query that should have created or updated a user in the `upsert`? If create, can you share the payload that caused this? If update, can you share the existing data and what you tried to update it to?

This looks to me like there was some kind of data inconsistency that lead to an empty result being returned which was unexpected by our Query Engine.

**Comment by khufu-adecoster** [comment]
Hi @janpio and thank you for your answer 🙏

Oh yes of course sorry I forgot to say it
It's a common occurrence
This error occurred with the **update**
The **create** works as expected
The **update** was executed without prior modification of the data created with the **create** of the `upsert`
This is a screenshot of mongodb data generated with the **create** of the `upsert`

<img width="398" alt="Capture d’écran 2023-08-04 à 17 47 18" src="https://github.com/prisma/prisma/assets/56578590/d8d5fe1c-27cc-4e41-8fd4-3f54e1080d96">

**Comment by janpio** [comment]
Interesting, and concerning. 

Can you maybe create a minimal reproduction in a new project that only uses the relevant part of the schema and the query you described? If this is a Prisma problem, it should be reproducible this way as well. (https://github.com/prisma/prisma-examples/tree/latest/typescript/script is usally a good starting point)

**Comment by khufu-adecoster** [comment]
Here it's ☺️

Same error message with the **schema** and **code** below
But I found some information that may be of interest. If I delete the `UsersCreatedBy` relation or the `UsersUpdatedBy` relation, it works perfectly

**Schema**
`generator client {
  provider        = "prisma-client-js"
  previewFeatures = []
}

datasource db {
  provider = "mongodb"
  url      = env("DATABASE_URL")
}

model Users {
  //ids
  id           String  @id @default(cuid()) @map("_id")
  //informations
  name         String?
  //create
  createdById  String?
  createdBy    Users?  @relation("UsersCreatedBy", fields: [createdById], references: [id], onDelete: NoAction, onUpdate: NoAction)
  //update
  updatedById  String?
  updatedBy    Users?  @relation("UsersUpdatedBy", fields: [updatedById], references: [id], onDelete: NoAction, onUpdate: NoAction)
  //fk users
  createdUsers Users[] @relation("UsersCreatedBy")
  updatedUsers Users[] @relation("UsersUpdatedBy")
}
`

**Code**
`import { PrismaClient } from '@prisma/client'

const prisma = new PrismaClient()

const name = 'toto'

prisma.users.upsert({
  where: {
    id: 'clkwidqx90000ygslp5myppfj'
  },
  create: {
    id: 'clkwidqx90000ygslp5myppfj',
    name
  },
  update: {
    name
  }
})
  .then(res => console.log('res', res))
  .catch(err => console.log('err', err))`

**Comment by SevInf** [comment]
Can confirm the issue by running code from https://github.com/prisma/prisma/issues/20527#issuecomment-1665474907 twice. Can also confirm that when removing `UsersCreatedBy` and `UsersUpdatedBy` panic goes away.

**Comment by janpio** [comment]
@khufu-adecoster Are you using `relationMode=prisma` in your app?

**Comment by khufu-adecoster** [comment]
@janpio No actually I didn't had that option
With this option I obtain a **Attempted to serialize empty result**

**Code**
`prisma.users.upsert({
        where: {
          firebaseId: uid
        },
        update: {
          email,
          phone: phoneNumber
        },
        create: {
          firebaseId: uid,
          email,
          name
        },
        include: {
          companies: true
        }
      })`

**Error**
`thread 'tokio-runtime-worker' panicked at 'Internal error: Attempted to serialize empty result.', query-engine/core/src/response_ir/ir_serializer.rs:57:40
stack backtrace:
   0:        0x13719d3bb - _napi_register_module_v1
   1:        0x1369a69cb - <unknown>
   2:        0x137176e5e - _napi_register_module_v1
   3:        0x1371a16da - _napi_register_module_v1
   4:        0x1371a12d5 - _napi_register_module_v1
   5:        0x1371a22a8 - _napi_register_module_v1
   6:        0x1371a1d77 - _napi_register_module_v1
   7:        0x1371a1d09 - _napi_register_module_v1
   8:        0x1371a1cf2 - _napi_register_module_v1
   9:        0x1374bdd23 - _napi_register_module_v1
  10:        0x136888986 - <unknown>
  11:        0x136889112 - <unknown>
  12:        0x136889bf4 - <unknown>
  13:        0x1368a723a - <unknown>
  14:        0x1368a3975 - <unknown>
  15:        0x136890d17 - <unknown>
  16:        0x1368f92ad - <unknown>
  17:        0x1368c3dfb - <unknown>
  18:        0x1368d2942 - <unknown>
  19:        0x1371b9ca8 - _napi_register_module_v1
  20:        0x1371c6421 - _napi_register_module_v1
  21:        0x1371be916 - _napi_register_module_v1
  22:        0x1371be3bc - _napi_register_module_v1
  23:        0x1371bfa2f - _napi_register_module_v1
  24:        0x1371a3aab - _napi_register_module_v1
  25:     0x7ff8120901d3 - __pthread_start
Err 1001 PrismaClientRustPanicError: 
Invalid `prisma.users.upsert()` invocation in
/Users/adecoster/code/projects/clokizi/api/src/middlewares/auth.ts:28:39

  25 //   }
  26 // })
  27 
→ 28 const user = await prisma.users.upsert(
Internal error: Attempted to serialize empty result.

This is a non-recoverable error which probably happens when the Prisma Query Engine has a panic.`

**Comment by janpio** [comment]
Oh of course, MongoDB does use the `relationMode=prisma` under the hood implicitly. Sorry about that.

**Comment by laplab** [comment]
Thank you for reporting the issue, @khufu-adecoster! We are currently testing the fix for this problem. Would it be possible for you to check your code against this version of Prisma?
```
npm i prisma@5.13.0-dev.30
```
Bear in mind that this is a development version and should not be used in production. This is only to check if the issue is fixed.

**Comment by khufu-adecoster** [comment]
Hey @laplab,
Sorry for the late reply 🙏
I confirm that on my test project for this issue, the problem is resolved with these versions ✅ 🚀

**Comment by laplab** [comment]
This issue was fixed in [Prisma 5.13.0](https://github.com/prisma/prisma/releases/tag/5.13.0).
